File: 0104-Maximum_Depth_of_Binary_Tree/maximum_depth.go
package maximumdepthofbinarytree

import "github.com/tjchiang/leetcode-exercise/structures"

type TreeNode = structures.TreeNode

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func maxDepthRecursion(root *TreeNode) int {
	if root == nil {
		return 0
	}
	leftDepth, rightDepth := maxDepthRecursion(root.Left), maxDepthRecursion(root.Right)

	// + 1 是當前中間節點
	return max(leftDepth, rightDepth) + 1
}

func maxDepthIteration(root *TreeNode) int {
	result := 0
	queue := make([]*TreeNode, 0)
	if root != nil {
		queue = append(queue, root)
	}

	// level order traversal
	for len(queue) > 0 {
		s := len(queue)
		for i := 0; i < s; i++ {
			node := queue[0]
			if node.Left != nil {
				queue = append(queue, node.Left)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
			}
			queue = queue[1:]
		}
		result++
	}

	return result
}

File: leetcode/0104-Maximum_Depth_of_Binary_Tree/maximum_depth.cpp
#include <iostream>
#include <vector>
#include <queue>
#include "../structures/binary_tree.h"
#include "../structures/tree_node.h"

using namespace std;

class Solution {
public:
    int maxDepth(TreeNode* root) {
        if (root == nullptr) return 0;
        const int left_max = maxDepth(root->left);
        const int right_max = maxDepth(root->right);

        return max(left_max, right_max) + 1;
    }

    int maxDepthIteration(TreeNode* root) {
        if (root == nullptr) return 0;
        int result = 0;
        queue<TreeNode*> node_queue;
        node_queue.push(root);

        while (!node_queue.empty()) {
            const int length = node_queue.size();
            result++;
            for (int i = 0; i < length; i++) {
                TreeNode* cur = node_queue.front();
                node_queue.pop();
                if (cur->left != nullptr) node_queue.push(cur->left);
                if (cur->right != nullptr) node_queue.push(cur->right);
            }
        }

        return result;
    }
};

int main(int argc, char* argv[])
{
    BinaryTree bt1(vector<int>{3,9,20,INT_NULL_,INT_NULL_,15,7});
    BinaryTree bt2(vector<int>{1,INT_NULL_,2});
    Solution sol;
    cout << "Result1: " << sol.maxDepth(bt1.get_root()) << endl;
    cout << "Result2: " << sol.maxDepth(bt2.get_root()) << endl;

    return 0;
}
